python/pyspark/sql/functions.py

```diff
@@ -91,13 +92,48 @@ def lit(col):
     Creates a :class:`Column` of literal value.
 
     .. versionadded:: 1.3.0
+    .. versionchanged:: 3.2.0
+        Added support for complex type literals.
 
     Examples
     --------
     >>> df.select(lit(5).alias('height')).withColumn('spark_user', lit(True)).take(1)
     [Row(height=5, spark_user=True)]
-    """
-    return col if isinstance(col, Column) else _invoke_function("lit", col)
+    >>> df.select(
+    ...     lit({"height": 5}).alias("data"),
+    ...     lit(["python", "scala"]).alias("languages")
+    ... ).take(1)
+    [Row(data={'height': 5}, languages=['python', 'scala'])]
+    """
+    if isinstance(col, Column):
+        return col
+
+    elif isinstance(col, list):
+        return array(*[lit(x) for x in col])
+
+    elif isinstance(col, tuple):
+        fields = (
+            # Named tuple
+            col._fields if hasattr(col, "_fields")
+            # PySpark Row
+            else col.__fields__ if hasattr(col, "__fields__")
+            # Other
+            else [f"_{i + 1}" for i in range(len(col))]
+        )
+
+        return struct(*[
```

Comment:
One thing that makes me worry is this isn't technically a literal ... although it's foldable, the results and optimization might be differently applied compared to Scala side .. also looks like Scala side doesn't support map and struct (due to both the limitation of typing and the fact that `Row` doesn't hold field names, etc. like PySpark).

Reply:
Hm, okay but these will be literals after constant folding though. @cloud-fan, WDYT about supporting complex type at literals in this way? I think we can do similar thing in Scala side as well to allow map and struct literals.

Reply:
cc @viirya too

Reply:
Seems okay because there is constant folding optimization. So if they are foldable, they will be optimized as literals during query optimization.

Reply:
> Also looks like Scala side doesn't support map and struct (due to both the limitation of typing and the fact that `Row` doesn't hold field names, etc.).

Actually, `typedLit` supports `Map` objects:

```scala
scala> typedLit(Map("a" -> 1, "b" -> 2))
res7: org.apache.spark.sql.Column = keys: [a,b], values: [1,2]
```

and `Products`

```scala
scala> case class FooBar(foo: Int, bar: String)
defined class FooBar

scala> typedLit(Foo(1, "a"))
res8: org.apache.spark.sql.Column = [1,a]
```

`Row` would be specific to PySpark, but since it is just a tuple, it seemed reasonable to handle it with names.

> Hm, okay but these will be literals after constant folding though.

The biggest difference compared to Scala is that it we handle mixed constants and column objects:

```python
>>> spark.createDataFrame([(1, 2)], ("a", "b")).select(lit({"a": col("a"), "b": col("b")})).show()      
+----------------+
| map(a, a, b, b)|
+----------------+
|{a -> 1, b -> 2}|
+----------------+
```

So there is a case where `lit` results in a non-`Literal`.

Reply:
Looks fine, too. As @zero323 said above, `typedLit` supports the complex types.

Reply:
It seems fine, but we might be able to convert these exprs for complex types into literals in advance via `PythonSQLUtils`.